app/uk/gov/hmrc/emailverification/controllers/EmailPasscodeController.scala

```diff
@@ -0,0 +1,118 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.emailverification.controllers
+
+import config.AppConfig
+import javax.inject.Inject
+import play.api.Logger
+import play.api.libs.json.{JsError, JsSuccess, JsValue, Json}
+import play.api.mvc._
+import uk.gov.hmrc.emailverification.connectors.{EmailConnector, PlatformAnalyticsConnector}
+import uk.gov.hmrc.emailverification.models._
+import uk.gov.hmrc.emailverification.repositories.{PasscodeMongoRepository, VerifiedEmailMongoRepository}
+import uk.gov.hmrc.http.{HeaderCarrier, UpstreamErrorResponse}
+import uk.gov.hmrc.play.audit.AuditExtensions._
+import uk.gov.hmrc.play.audit.http.connector.AuditConnector
+import uk.gov.hmrc.play.audit.model.ExtendedDataEvent
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Random
+
+
+class EmailPasscodeController @Inject()(emailConnector: EmailConnector,
+                                        passcodeRepo:PasscodeMongoRepository,
+                                        verifiedEmailRepo: VerifiedEmailMongoRepository,
+                                        analyticsConnector: PlatformAnalyticsConnector,
+                                        auditConnector: AuditConnector,
+                                        controllerComponents: ControllerComponents
+                                            )(implicit ec:ExecutionContext, appConfig: AppConfig) extends BaseControllerWithJsonErrorHandling(controllerComponents) {
+
+  def newPasscode(): String =  {
+    val codeSize = 6
+    Random.alphanumeric
+      .filterNot(_.isDigit)
+      .filterNot(_.isLower)
+      .filterNot(Set('A', 'E', 'I', 'O', 'U'))
+      .take(codeSize).mkString
+  }
+
+  private def sendEmailAndStorePasscode(request: PasscodeRequest)(implicit hc: HeaderCarrier) = {
+    val passcode = newPasscode()
+    val paramsWithPasscode = request.templateParameters.getOrElse(Map.empty) +
+      ("passcode" -> passcode)
+
+    for {
+      _ <- emailConnector.sendEmail(request.email, request.templateId, paramsWithPasscode)
+      _ <- passcodeRepo.upsert(request.sessionId, passcode, request.email, request.linkExpiryDuration)
+    } yield Created
+  }
+
+  def requestPasscode(): Action[JsValue] = Action.async(parse.json) {
+    implicit httpRequest =>
+      withJsonBody[PasscodeRequest] { request =>
+        verifiedEmailRepo.isVerified(request.email) flatMap {
+          case true => Future.successful(Conflict(Json.toJson(ErrorResponse("EMAIL_VERIFIED_ALREADY", "Email has already been verified"))))
+          case false =>
+            analyticsConnector.sendEvents(GaEvents.passcodeRequested)
+            sendEmailAndStorePasscode(request)
```

Comment:
I'd personally prefer e.g. `sendEmailAndStorePasscode(request).map(_ => Created)` (and sendEmailAndStorePasscode returning `Future[Unit]`) to make it clearer what response codes `requestPasscode` might return

Reply:
yep sounds good